docs/onion-message-channels.md

```diff
@@ -0,0 +1,220 @@
+
+### Contents
+
+1. [Overview](#overview)
+
+2. [Testing, configuring for signet](#testing)
+
+4. [Directory nodes](#directory)
+
+<a name="overview" />
+
+## Overview
+
+This is a new way for Joinmarket bots to communicate, namely by serving and connecting to Tor onion services. This does not
+introduce any new requirements to your Joinmarket installation, technically, because the use of Payjoin already required the need
+to service such onion services, and connecting to IRC used a SOCKS5 proxy (by default, and used by almost all users) over Tor to
+a remote onion service.
+
+The purpose of this new type of message channel is as follows:
+
+* less reliance on any service external to Joinmarket
+* most of the transaction negotiation will be happening directly peer to peer, not passed over a central server (
+albeit it was and remains E2E encrypted data, in either case)
+* the above can lead to better scalability at large numbers
+* a substantial increase in the speed of transaction negotiation; this is mostly related to the throttling of high bursts of traffic on IRC
+
+The configuration for a user is simple; in their `joinmarket.cfg` they will get a messaging section like this, if they start from scratch:
+
+```
+[MESSAGING:onion1]
+# onion based message channels must have the exact type 'onion'
+# (while the section name above can be MESSAGING:whatever), and there must
+# be only ONE such message channel configured (note the directory servers
+# can be multiple, below):
+type = onion
+
+socks5_host = localhost
+socks5_port = 9050
+
+# the tor control configuration.
+# for most people running the tor daemon
+# on Linux, no changes are required here:
+tor_control_host = localhost
+# or, to use a UNIX socket
+# tor_control_host = unix:/var/run/tor/control
+tor_control_port = 9051
+
+# the host/port actually serving the hidden service
+# (note the *virtual port*, that the client uses,
+# is hardcoded to 80):
+onion_serving_host = 127.0.0.1
+onion_serving_port = 8080
+
+# directory node configuration
+#
+# This is mandatory for directory nodes (who must also set their
+# own *.onion:port as the only directory in directory_nodes, below),
+# but NOT TO BE USED by non-directory nodes (which is you, unless
+# you know otherwise!), as it will greatly degrade your privacy.
+# (note the default is no value, don't replace it with "").
+hidden_service_dir =
+#
+# This is a comma separated list (comma can be omitted if only one item).
+# Each item has format host:port ; both are required, though port will
+# be 80 if created in this code.
+directory_nodes = rr6f6qtleiiwic45bby4zwmiwjrj3jsbmcvutwpqxjziaydjydkk5iad.onion:80
+
+# This setting is ONLY for developer regtest setups,
+# running multiple bots at once. Don't alter it otherwise
+regtest_count = 0,0
+```
+
+All of these can be left as default for most users, except the field `directory_nodes`.
+
+The list of **directory nodes** (the one shown here is one being run on signet, right now), which will
```

Comment:
Shouldn't we have different directory node lists for mainnet / signet / testnet? I like being able to switch networks with the same joinmarket.cfg with minimum number of lines needed to be changed.

Reply:
Yes it's probably time to do that. Except not testnet3, I would never use that. I can set one up for mainnet now (it's probably best for it to be me for now, because I'll know when an urgent update is needed. We should edit it to be something sensible before release, though, of course).
(So TLDR action item: I will set up one mainnet directory node shortly and add it here).

Reply:
OK reviewing this again: I see what you mean, it will be (slightly) annoying to re-configure directory nodes for different networks, and it will be annoying to run multiple directory nodes at once for different networks, instead of just having the logical equivalent of "channels" as in IRC, so that you just naturally talk to peers of the right type by choosing the right channel. (Also very confusing is that I used `MessageChannel`, several years back, to define a thing inside Joinmarket itself, which is more like 'server' in this context, than 'channel'! And then there's the fact we are now using a p2p element as well ...).

Since I think this will take some careful thought, my intuition is to leave it alone for now; I am worried about adding more complexity.

The handshake has two deliberately forwards-compatible elements, namely the `features` dict, which can be filled with absolutely anything, combined with the `version` range check, so a future version can be supported concurrently in a running directory node. And the future version could take data from the `features` dict to figure out how to handle a specific peer (e.g. allocate them to signet, not mainnet, I guess).

For now we'll just run signet vs mainnet d-nodes as separate bots, and users can have two different directory-node lists in their config, and comment out as appropriate. (and ofc most users will just use mainnet anyway, so they won't care about this).

Reply:
As of https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1182/commits/755b508da374af731b12ec0bfcc416660533a93d I think this is resolved satisfactorily, at least for now.

Bots will announce their configured network (`[BLOCKCHAIN]` field `network`) to anyone they handshake to, and it must match for the handshake to be accepted, else the connection is just dropped. I've also updated the docs to reflect this point, including advice to directory node runners to be sure to update that field in their config (because they don't have to update the *other* fields in BLOCKCHAIN, as they are all unused).

This is not 100% the best, but imo it's definitely good enough.